File: 1122-relative-sort-array/frequency-map-based-relative-sort/cpp/solution.cpp
class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        map<int, int> freq_map;
        for (int num : arr1) 
            freq_map[num]++;

        vector<int> res;

        // Add elements in the order defined by arr2, and remove from the frequency map
        for (int num : arr2) {
            if (freq_map.count(num)) {
                int count = freq_map[num];
                res.insert(res.end(), count, num);
                freq_map.erase(num);  
            }
        }

        // Add remaining elements in ascending order
        for (auto const& [val, count] : freq_map) {
            res.insert(res.end(), count, val);
        }
        return res;
    }
};

File: 1122-relative-sort-array/frequency-map-based-relative-sort/java/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        TreeMap<Integer, Integer> freqMap = new TreeMap<>();
        for (int num : arr1) 
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);

        int[] res = new int[arr1.length];
        int idx = 0;

        // Add elements in the order defined by arr2, and remove from the frequency map
        for (int num : arr2) {
            Integer count = freqMap.remove(num);
            if (count != null) {
                while (count-- > 0) 
                    res[idx++] = num;
            }
        }

        // Add remaining elements in ascending order
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            int count = entry.getValue();
            int val = entry.getKey();
            for (int i = 0; i < count; i++) {
                res[idx++] = val;
            }
        }
        return res;
    }
}
